File: KeeperLegacyTests/ProgressionTests.swift
import XCTest
@testable import KeeperLegacy

// MARK: - Progression Tests
// Tests for XP curve, leveling, feature unlocks, and milestone rewards.

final class ProgressionTests: XCTestCase {

    // MARK: XP Curve

    func testXPRequiredIsPositive() {
        for level in 1...49 {
            XCTAssertGreaterThan(
                XPCurve.xpRequired(forLevel: level), 0,
                "Level \(level) requires 0 XP — invalid"
            )
        }
    }

    func testEarlyLevelsHaveLowerXPThanLateOnes() {
        let early = XPCurve.xpRequired(forLevel: 3)
        let late  = XPCurve.xpRequired(forLevel: 30)
        XCTAssertLessThan(early, late)
    }

    // MARK: Leveling Up

    func testAddXPTriggersLevelUp() {
        var progression = PlayerProgression()
        XCTAssertEqual(progression.currentLevel, 1)
        let xpNeeded = XPCurve.xpRequired(forLevel: 1)
        let levelsGained = progression.addXP(xpNeeded)
        XCTAssertEqual(levelsGained, [2])
        XCTAssertEqual(progression.currentLevel, 2)
    }

    func testXPDoesNotExceedMaxLevel() {
        var progression = PlayerProgression()
        progression.currentLevel = 50
        progression.currentXP    = 0
        progression.addXP(999999)
        XCTAssertEqual(progression.currentLevel, 50)
        XCTAssertEqual(progression.currentXP, 0)  // Resets at max
    }

    func testXPRemainsAfterLevelUp() {
        var progression = PlayerProgression()
        let xpNeeded = XPCurve.xpRequired(forLevel: 1)
        progression.addXP(xpNeeded + 25)
        XCTAssertEqual(progression.currentLevel, 2)
        XCTAssertEqual(progression.currentXP, 25)  // 25 carries over
    }

    // MARK: Feature Unlocks

    func testBasicFeaturesUnlockedAtStart() {
        let progression = PlayerProgression()
        XCTAssertTrue(progression.isFeatureUnlocked(.shop))
        XCTAssertTrue(progression.isFeatureUnlocked(.habitat))
        XCTAssertTrue(progression.isFeatureUnlocked(.feeding))
        XCTAssertTrue(progression.isFeatureUnlocked(.playing))
    }

    func testBreedingLockedAtStart() {
        let progression = PlayerProgression()
        XCTAssertFalse(progression.isFeatureUnlocked(.breeding))
    }

    func testBreedingUnlocksAtLevel12WithAct1() {
        var progression = PlayerProgression()
        progression.currentLevel = 12
        progression.storyAct     = 1      // Act I complete
        progression.checkAndUnlockFeatures()
        XCTAssertTrue(progression.isFeatureUnlocked(.breeding))
    }

    func testBreedingRemainsLockedAtLevel12WithoutStory() {
        var progression = PlayerProgression()
        progression.currentLevel = 12
        progression.storyAct     = 0      // Act I NOT complete
        progression.checkAndUnlockFeatures()
        XCTAssertFalse(progression.isFeatureUnlocked(.breeding))
    }

    func testMagicalHabitatRequiresAct2() {
        var progression = PlayerProgression()
        progression.currentLevel = 50
        progression.storyAct     = 1
        progression.checkAndUnlockFeatures()
        XCTAssertFalse(progression.isFeatureUnlocked(.magicalHabitat))

        progression.advanceStoryAct(to: 2)
        XCTAssertTrue(progression.isFeatureUnlocked(.magicalHabitat))
    }

    // MARK: Milestones

    func testMilestoneExistsAtLevel5() {
        XCTAssertNotNil(MilestoneReward.milestones[5])
    }

    func testMilestoneLevel10HasStoryEvent() {
        let milestone = MilestoneReward.milestones[10]
        XCTAssertNotNil(milestone?.storyEvent)
    }

    func testMilestoneLevel50HasLargeBonus() {
        let milestone = MilestoneReward.milestones[50]
        XCTAssertNotNil(milestone)
        XCTAssertGreaterThan(milestone!.coinBonus, 1000)
    }
}

File: KeeperLegacyGodot/Tests/ProgressionTests.cs
// Tests/ProgressionTests.cs
// Port of ProgressionTests.swift

using NUnit.Framework;
using KeeperLegacy.Models;

namespace KeeperLegacy.Tests
{
    [TestFixture]
    public class ProgressionTests
    {
        // ── XP Curve ──────────────────────────────────────────────────────────

        [Test]
        public void XPRequiredIsPositive()
        {
            for (int level = 1; level <= 49; level++)
                Assert.That(XPCurve.XpRequired(level), Is.GreaterThan(0),
                    $"Level {level} requires 0 XP — invalid");
        }

        [Test]
        public void EarlyLevelsHaveLowerXPThanLateOnes()
        {
            int early = XPCurve.XpRequired(3);
            int late  = XPCurve.XpRequired(30);
            Assert.That(early, Is.LessThan(late));
        }

        // ── Leveling Up ───────────────────────────────────────────────────────

        [Test]
        public void AddXPTriggersLevelUp()
        {
            var p = new PlayerProgression();
            Assert.That(p.CurrentLevel, Is.EqualTo(1));
            int xpNeeded = XPCurve.XpRequired(1);
            var levels = p.AddXP(xpNeeded);
            Assert.That(levels, Is.EqualTo(new[] { 2 }));
            Assert.That(p.CurrentLevel, Is.EqualTo(2));
        }

        [Test]
        public void XPDoesNotExceedMaxLevel()
        {
            var p = new PlayerProgression();
            p.CurrentLevel = 50;
            p.CurrentXP    = 0;
            p.AddXP(999999);
            Assert.That(p.CurrentLevel, Is.EqualTo(50));
            Assert.That(p.CurrentXP,    Is.EqualTo(0));
        }

        [Test]
        public void XPRemainsAfterLevelUp()
        {
            var p = new PlayerProgression();
            int xpNeeded = XPCurve.XpRequired(1);
            p.AddXP(xpNeeded + 25);
            Assert.That(p.CurrentLevel, Is.EqualTo(2));
            Assert.That(p.CurrentXP,    Is.EqualTo(25));
        }

        // ── Feature Unlocks ───────────────────────────────────────────────────

        [Test]
        public void BasicFeaturesUnlockedAtStart()
        {
            var p = new PlayerProgression();
            Assert.That(p.IsFeatureUnlocked(GameFeature.Shop),    Is.True);
            Assert.That(p.IsFeatureUnlocked(GameFeature.Habitat),  Is.True);
            Assert.That(p.IsFeatureUnlocked(GameFeature.Feeding),  Is.True);
            Assert.That(p.IsFeatureUnlocked(GameFeature.Playing),  Is.True);
        }

        [Test]
        public void BreedingLockedAtStart()
        {
            var p = new PlayerProgression();
            Assert.That(p.IsFeatureUnlocked(GameFeature.Breeding), Is.False);
        }

        [Test]
        public void BreedingUnlocksAtLevel12WithAct1()
        {
            var p = new PlayerProgression();
            p.CurrentLevel = 12;
            p.StoryAct     = 1;
            p.CheckAndUnlockFeatures();
            Assert.That(p.IsFeatureUnlocked(GameFeature.Breeding), Is.True);
        }

        [Test]
        public void BreedingRemainsLockedAtLevel12WithoutStory()
        {
            var p = new PlayerProgression();
            p.CurrentLevel = 12;
            p.StoryAct     = 0;   // Act I not complete
            p.CheckAndUnlockFeatures();
            Assert.That(p.IsFeatureUnlocked(GameFeature.Breeding), Is.False);
        }

        [Test]
        public void MagicalHabitatRequiresAct2()
        {
            var p = new PlayerProgression();
            p.CurrentLevel = 50;
            p.StoryAct     = 1;
            p.CheckAndUnlockFeatures();
            Assert.That(p.IsFeatureUnlocked(GameFeature.MagicalHabitat), Is.False);

            p.AdvanceStoryAct(2);
            Assert.That(p.IsFeatureUnlocked(GameFeature.MagicalHabitat), Is.True);
        }

        // ── Milestones ────────────────────────────────────────────────────────

        [Test] public void MilestoneExistsAtLevel5()
            => Assert.That(MilestoneReward.Milestones.ContainsKey(5), Is.True);

        [Test]
        public void MilestoneLevel10HasStoryEvent()
        {
            Assert.That(MilestoneReward.Milestones.TryGetValue(10, out var m), Is.True);
            Assert.That(m!.StoryEvent, Is.Not.Null);
        }

        [Test]
        public void MilestoneLevel50HasLargeBonus()
        {
            Assert.That(MilestoneReward.Milestones.TryGetValue(50, out var m), Is.True);
            Assert.That(m!.CoinBonus, Is.GreaterThan(1000));
        }
    }
}
